src/natcap/invest/sdr/sdr_core.pyx

```diff
@@ -557,4 +567,136 @@ def calculate_sediment_deposition(
+        mfd_flow_direction_path, 1, False)
+
+    cdef _ManagedRaster average_aspect_raster = _ManagedRaster(
+        target_average_aspect_path, 1, True)
+
+    cdef int seed_row = 0
+    cdef int seed_col = 0
+    cdef int n_pixels_visited = 0
+    cdef int win_xsize, win_ysize, xoff, yoff
+    cdef int row_index, col_index, neighbor_index
+    cdef int flow_weight_in_direction
+    cdef int weight_sum
+    cdef int seed_flow_value
+    cdef float aspect_weighted_average, aspect_weighted_sum
+    cdef float proportional_flow
+    cdef float flow_length
+
+    # the flow_lengths array is the functional equivalent
+    # of calculating |sin(alpha)| + |cos(alpha)|.
+    cdef float* flow_lengths = [
+        1.0, <float>cmath.M_SQRT2,
+        1.0, <float>cmath.M_SQRT2,
+        1.0, <float>cmath.M_SQRT2,
+        1.0, <float>cmath.M_SQRT2
+    ]
+
+    # Loop over iterblocks to maintain cache locality
+    # Find each non-nodata pixel and calculate proportional flow
+    # Multiply proportional flow times the flow length x_d
+    # write the final value to the raster.
+    for offset_dict in pygeoprocessing.iterblocks(
+            (mfd_flow_direction_path, 1), offset_only=True, largest_block=0):
+        win_xsize = offset_dict['win_xsize']
+        win_ysize = offset_dict['win_ysize']
+        xoff = offset_dict['xoff']
+        yoff = offset_dict['yoff']
+
+        LOGGER.info('Average aspect %.2f%% complete', 100.0 * (
+            n_pixels_visited / float(n_cols * n_rows)))
+
+        for row_index in range(win_ysize):
+            seed_row = yoff + row_index
+            for col_index in range(win_xsize):
+                seed_col = xoff + col_index
+                seed_flow_value = <int>mfd_flow_direction_raster.get(
+                    seed_col, seed_row)
+
+                # Skip this seed if it's nodata (Currently expected to be 0).
+                # No need to set the nodata value here since we have already
+                # filled the raster with nodata values at creation time.
+                if seed_flow_value == mfd_flow_direction_nodata:
+                    continue
+
+                weight_sum = 0
+                aspect_weighted_sum = 0
+                for neighbor_index in range(8):
+                    # We only need to check the neighbors relative to the
+                    # boundaries if we're in the first or last rows and
+                    # columns.
+                    if seed_row == 0 or seed_row == n_rows:
+                        neighbor_row = seed_row + ROW_OFFSETS[neighbor_index]
+                        if neighbor_row < 0 or neighbor_row >= n_rows:
+                            continue
+
+                    if seed_col == 0 or seed_col == n_cols:
+                        neighbor_col = seed_col + ROW_OFFSETS[neighbor_index]
+                        if neighbor_col < 0 or neighbor_col >= n_cols:
+                            continue
+
```

Comment:
Can the whole section above be:
```
neighbor_col = seed_col + COL_OFFSETS[neighbor_index]
if neighbor_col < 0 or neighbor_col == n_cols:
    continue
neighbor_row = seed_row + ROW_OFFSETS[neighbor_index]
if neighbor_row < 0 or neighbor_row == n_rows:
    continue
```

Reply:
Of course it can be (and it was originally).  The additional checking came out of profiling the file line-by-line.  Having the extra conditional (`if seed_col == 0 or seed_col == n_cols:`), while a bit uglier, did seem to reduce the overall runtime.

So, I timed this out!  Here are the mean times for these different implementations running on the pit-filled (via pygeoprocessing) Willamette DEM.  Tests were run over 100 repetitions.  Times reported are the mean time per call to `sdr_core.calculate_average_aspect`:

```python
# Implementation 1: This is the implementation I had in the PR.
# Mean time per repetition: 0.32161s
if seed_row == 0 or seed_row == (n_rows -1):
    neighbor_row = seed_row + ROW_OFFSETS[neighbor_index]
    if neighbor_row < 0 or neighbor_row >= n_rows:
        continue
if seed_col == 0 or seed_col == (n_cols -1):
    neighbor_col = seed_col + COL_OFFSETS[neighbor_index]
    if neighbor_col < 0 or neighbor_col >= n_cols:
        continue
```

```python
# Implementation 2: This allows us to test whether the initial seed checks are really worth it.
# Mean time per repetition: 0.32169s
if seed_row == 0 or seed_row == (n_rows -1):
    neighbor_row = seed_row + ROW_OFFSETS[neighbor_index]
    if neighbor_row == -1 or neighbor_row == n_rows:
        continue
if seed_col == 0 or seed_col == (n_cols -1):
    neighbor_col = seed_col + COL_OFFSETS[neighbor_index]
    if neighbor_col == -1 or neighbor_col == n_cols:
        continue
```

```python
# Implementation 3: This is what was requested by @richpsharp 
# Mean time per repetition: 0.327s
neighbor_col = seed_col + COL_OFFSETS[neighbor_index]
if neighbor_col < 0 or neighbor_col == n_cols:
    continue
neighbor_row = seed_row + ROW_OFFSETS[neighbor_index]
if neighbor_row < 0 or neighbor_row == n_rows:
    continue
```

```python
# Implementation 4: Taking @richpsharp 's suggestion a step further
# Mean time per repetition: 0.32152s
neighbor_col = seed_col + COL_OFFSETS[neighbor_index]
if neighbor_col == -1 or neighbor_col == n_cols:
    continue
neighbor_row = seed_col + ROW_OFFSETS[neighbor_index]
if neighbor_row == -1 or neighbor_row == n_rows:
    continue
```
Interestingly, while the line-profiling might have indicated that implementations 1 or 2 should have been faster, they were not.  I would have expected implementations 2 and 4 to have been about the same runtime based on what line-profiler was telling me, but the timings here would indicate that this routine is fastest when we're only checking for exactly certain row/col indexes in the most concise way.

So, implementation 4 is what's included in 8ac25750.